alibi/explainers/similarity/grad.py

```diff
@@ -5,82 +5,139 @@
+    REGRESSION = "regression"
+
+    @staticmethod
+    def from_str(name: str):
+        return {
+            'classification': Task.CLASSIFICATION,
+            'regression': Task.REGRESSION
+        }[name]
+
+    @staticmethod
+    def values():
+        return {item.value for item in Task.__members__.values()}
+
+    @staticmethod
+    def options_string():
+        return f"""'{"' | '".join(Task.values())}'"""
+
+
+class GradientSimilarity(BaseSimilarityExplainer):
+
     def __init__(self,
                  predictor: 'Union[tensorflow.keras.Model, torch.nn.Module]',
-                 loss_fn: '''Callable[[Union[tensorflow.Tensor, torch.Tensor],
-                                    Union[tensorflow.Tensor, torch.Tensor]],
-                                   Union[tensorflow.Tensor, torch.Tensor]]''',
-                 sim_fn: 'Union[Callable[[np.ndarray, np.ndarray], np.ndarray], str]' = 'grad_dot',
-                 task: str = "classification",
+                 loss_fn: '''Union[Callable[[tensorflow.Tensor, tensorflow.Tensor], tensorflow.Tensor],
+                                   Callable[[torch.Tensor, torch.Tensor], torch.Tensor]]''',
+                 sim_fn: Literal['grad_dot', 'grad_cos', 'grad_asym_dot'] = 'grad_dot',
+                 task: Literal["classification", 'regression'] = 'classification',
                  store_grads: bool = False,
-                 seed: int = 0,
-                 backend: str = "tensorflow",
-                 **kwargs
+                 backend: Literal['tensorflow', 'pytorch'] = "tensorflow",
+                 device: 'Union[int, str, torch.device, None]' = None,
                  ):
-        """Constructor
+        """Constructor for GradientSimilarity explainer.
 
         Parameters
         ----------
-        predictor:
-            model to explain.
-        loss_fn:
-            Loss function used.
-        sim_fn:
-            Similarity function to use. ``'grad_dot'`` | ``'grad_cos'``. Default: ``'grad_dot'``.
-        task:
-            Task to perform. ``'classification'`` | ``'regression'``. Default: ``'classification'``.
-        store_grads:
-            Whether to store gradients. Default ``False``. If ``False``, gradients are computed on the fly otherwise we
-            store them which can be faster when it comes to explaining and instances.
-        seed:
-            Random seed. Default: 0.
-        backend:
-            Backend to use. ``'tensorflow'`` | ``'torch'``. Default: ``'tensorflow'``.
-        kwargs:
-            Additional arguments to pass to the similarity function.
+        predictor
+            Model to explain.
+        loss_fn
+            Loss function used. The gradient of the loss function is used to compute the similarity between the test
+            instances and the training set. This should be the same loss used to train the model.
+        sim_fn
+            Similarity function to use.
+        task
+            Task performed by the model.
```

Comment:
Should we explain what our assumptions are of the model given either of these task specifications? I.e. single-output for regression, potentially single-output (corresponding to label) or 2-D output (corresponding to probabilities) for classification?

Reply:
I think the only difference the choice of the task makes is that for regression the target needs to be specified explicitly whereas for classification it defaults to the model prediction. I've added documentation for that behaviour in the `explain` method. Should I document it here too?

This is all the `task` parameter does. Constraints on the shape of `Y` are only imposed by the `loss_function` and independent of the explainer itself.

Reply:
ok, perhaps worth documenting briefly here what this parameter does internally later on in the execution?